### PR TITLE
Support default properties for devices

### PR DESCRIPTION
**Pull request by sophokles73**
With this PR I would like to add support for registering *default* properties with a device which can be used by protocol adapters to augment messages published by the device with missing information.

One prominent example of such missing information is the *content-type*. A device publishing data via MQTT (currently) has no way of specifying a content type but in many cases the device will always publish the same type of content (e.g. a temperature reading or an Eclipse Unide measurement). In such cases, a default *content-type* (say `application/vnd.eclipse.unide-measurement+json`) can be registered with the device and the MQTT protocol adapter can set this content type on each downstream message from the device that does not already have a content type set.

We will be facing the same problem with any device that uses narrow band communication like sigfox, LoRaWAN etc. and for which the telemetry messages are being received by the particular *gateway* that the devices are connected to. Most of these gateways provide means to specify a URL where they can POST the data to (e.g. the Hono HTTP adapter) but do not support the specification of a content type.

In a subsequent PR I would like to add generic support for caching assertions and default properties in the protocol adapters so that for example the HTTP adapter does not need to invoke the *assert device registration* on the Device Registration service for each and every message being published.

### CODE REVIEW

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java [hunk 1]
```diff
@@ -416,6 +423,73 @@ protected final void closeClients(final Handler<AsyncResult<Void>> closeHandler)
...
+            if (defaults != null) {
+                addDefaults(message, defaults);
+            }
+        }
+        if (getConfig().isJmsVendorPropsEnabled()) {
+            MessageHelper.addJmsVendorProperties(message);
+        }
+    }
+
+    private void addDefaults(final Message message, final JsonObject defaults) {
```
Comment (ppatierno) [inline service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: missing summary for this method.

Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: it is a private method
so far, we haven't been adding JavaDocs to private methods IIRC.

**Comment by sophokles73** [comment]
> From the code it seems that even if the adapter has the "defaults" disabled, the device can still be registered sending the "defaults" in the JSON object; of course it means that even if the device registered with some "defaults", the adapter won't put them as properties because this feature is disabled. Do we need a way to have the device aware of that ? 

Aware of what?

> It can't be done on registration side because we could have two different adapters, one having "defaults" enabled but not the other one; in this case the registration is ok but noth both adapters will support the specified "defaults". Of course, doing something on the adapter side means that the adapter itself needs to inspect the payload searching for "defaults". Any thoughts about that ?

Sorry, I do not seem to understand the question. Can you be a little more specific about your concern(s)?

**Comment by ppatierno** [comment]
the device won't be aware that he put some "defaults" into the registration message but then these "defaults" won't be put inside the AMQP downstream messages because the protocol adapter it's connecting to has "defaults" disabled.

File: core/src/main/java/org/eclipse/hono/util/RegistrationConstants.java [hunk 2]
```diff
@@ -27,24 +25,51 @@
...
-    public static final String FIELD_ASSERTION                   = "assertion";
-    public static final String FIELD_DATA                        = "data";
-
+    /**
+     * The name of the field in a response to the <em>assert device registration</em> operation
+     * that contains the registration status assertion.
+     */
+    public static final String FIELD_ASSERTION    = "assertion";
+    /**
+     * The name of the field in a response to the <em>get</em> operation
```
Comment (sysexcontrol) [inline core/src/main/java/org/eclipse/hono/util/RegistrationConstants.java]: above the operation was described as <em>get registration operation</em>.

File: site/content/api/Device-Registration-API.md [hunk 3]
```diff
@@ -137,16 +137,20 @@ A response to an *assertion* request contains the [Standard Response Properties]
...
+| Name             | Mandatory | Type          | Description |
+| :--------------- | :-------: | :------------ | :---------- |
+| *device-id*      | *yes*     | *string*      | The ID of the device that is subject of the assertion. |
+| *assertion*      | *yes*     | *string*      | A [JSON Web Token](https://jwt.io/introduction/) which MUST contain the device id (`sub` claim), the tenant id (private `ten` claim) and an expiration time (`exp` claim). The token MAY contain additional claims as well. A client SHOULD silently ignore claims it does not understand. |
+| *defaults*       | *no*      | *JSON object* | Default values to be used by protocol adapters for augmenting messages from devices with missing information like a *content type*. It is up to the discretion of a protocol adapter if and how to use the given default values when processing messages published by the device. |
```
Comment (sysexcontrol) [inline site/content/api/Device-Registration-API.md]: Why shouldn't we make it mandatory for protocol adapters to use the "defaults"?
If defaults are not desired, they can just be not set in the device registry, but if we have some protocol adapters in the future that consider them, and some not, what do we gain?

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java [hunk 4]
```diff
@@ -387,24 +389,29 @@ protected final void closeClients(final Handler<AsyncResult<Void>> closeHandler)
...
 
     /**
-     * Gets a registration status assertion for a device.
+     * Gets an assertion for a device's registration status.
+     * <p>
+     * The returned JSON object contains the assertion for the device
+     * under property {@link RegistrationConstants#FIELD_ASSERTION}.
+     * <p>
+     * In addition to the assertion the returned object may include <em>default</em>
+     * values for properties to set on messages published by the device.
```
Comment (sysexcontrol) [inline service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: ... under property {@link RegistrationConstants.FIELD_DEFAULTS}. ???

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java [hunk 5]
```diff
@@ -496,25 +570,39 @@ protected final boolean validateCredentialsWithTopicStructure(final String tenan
...
+        return newMessage(address, deviceId, null);
+    }
+
     /**
      * Creates a new AMQP 1.0 message for an address, device ID and content type.
      * 
      * @param address The receiver of the message.
      * @param deviceId The identifier of the device that the message originates from.
-     * @param contentType The content type describing the message's payload.
+     * @param contentType The content type describing the message's payload (may be {@code null}.
```
Comment (sysexcontrol) [inline service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: ) missing

**Comment by sophokles73** [comment]
> the device won't be aware that he put some "defaults" into the registration message but then these "defaults" won't be put inside the AMQP downstream messages because the protocol adapter it's connecting to has "defaults" disabled.

I see. FMPOV it is usually **not** the device that registers *itself* but instead a device is usually registered as part of a provisioning process controlled by the organization or person that owns the device. I also think that the `defaultsEnabled` flag should (later) be configured at the *tenant* level, alongside the *authenticationReguired` flag etc.

For devices doing self-registration as you proposed, I do not have a convincing solution (for now ;-)).

File: site/content/api/Device-Registration-API.md (hunk 3 above)
Comment (sophokles73) [reply site/content/api/Device-Registration-API.md]: We can always make our protocol adapters use the default properties but I do not think that we can enforce this on custom adapters.

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java (hunk 4 above)
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: In JavaDoc tags, the `#` is the delimiter between class and field/method name.

File: site/content/api/Device-Registration-API.md (hunk 3 above)
Comment (sysexcontrol) [reply site/content/api/Device-Registration-API.md]: OK, I understand.
Additionally I haven't seen that defaults can be switched off (which makes sense to me), so I guess it should stay like it is...

**Comment by zubairhamed** [comment]
1. I think a "version" property could go nicely with the content type to differentiate versions of a device telemetry.
2. a lastseen attribute which shows the last telemetry/event emitted by a device
3. for vorto, maybe a "vorto" attribute pointing to a vorto model, if a consumer might want to use this for simplistic vorto transformations

**Comment by sophokles73** [comment]
> I think a "version" property could go nicely with the content type to differentiate versions of a device telemetry

If you are talking about the *version* of the content type, then there is nothing else we need to do. The content type can also contain a version property, e.g. `application/vnd.acme+json; version=2.0`

> a lastseen attribute which shows the last telemetry/event emitted by a device

Wouldn't the value be always *now*, given that it is a property being set on a telemetry/event message being published by the device?

> for vorto, maybe a "vorto" attribute pointing to a vorto model, if a consumer might want to use this for simplistic vorto transformations

The generic approach taken in this PR would not prevent you from registering such a default property for a device. In fact, you can register arbitrary properties that you want to be added to the downstream messages published by the device.

**Comment by sysexcontrol** [comment]
Regarding "lastseen":  
- I understood that it should be the value of the *previous* message, right?

But in general I have the question that so far I understand the default values to contain *static* values. However, this property already needs some *activity*, e.g. to remember the lastseen value somewhere (e.g. in the device registry). This IMHO contradicts the currently simple solution of default values addressed by this PR. 
Or did I fully misunderstand your proposal, @zubairhamed ?

**Pull request merged**